File: LeetCode/029 Divide 2 integers/Solution.java
public class Solution {
    public int divide(int dividend, int divisor) {
        long l_dividend = (long) dividend;
        long l_divisor = (long) divisor;

        int sign = 1;

        if (l_dividend < 0){
        	l_dividend = -l_dividend;
        	sign = -sign;
        }

        if (l_divisor < 0){
        	l_divisor = -l_divisor;
        	sign = -sign;
        }

        long quo = quotient(l_dividend, l_divisor);

        quo = quo*sign;

        if (quo > Integer.MAX_VALUE)
        	return Integer.MAX_VALUE;
        else if (quo < Integer.MIN_VALUE)
        	return Integer.MIN_VALUE;

        return (int)quo;

    }

    public long quotient(long dividend, long divisor){
    	long cnt = 1;
    	long orig = divisor;

    	while(divisor<<1 < dividend){
    		divisor <<=1;
    		cnt<<=1;
    	}

    	if (divisor == dividend){
    		return cnt;
    	}
    	else {
    		cnt = cnt + quotient(dividend - divisor, divisor);
    	}

    	return cnt;
    }
}

File: LeetCode/029 Divide 2 integers/Solution.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        quotient, dvd, dvs = 0, abs(dividend), abs(divisor)
        sign = (dividend < 0) is (divisor < 0)

        while dvd >= dvs:
            inc = dvs
            i = 0

            while dvd >= inc:
                dvd -= inc
                quotient += 1<<i
                i += 1
                inc<<=1

        if not sign:
            quotient = -quotient

        return min(max(-2147483648, quotient), 2147483647)
